rossum_api/api_client.py

```diff
@@ -175,23 +188,49 @@ async def fetch_all(
         )
         # Fire async tasks to fetch the rest of the pages and start yielding results from page 1
         last_page = min(total_pages, max_pages or total_pages)
-        page_requests = [
-            asyncio.create_task(
-                self._fetch_page(
-                    f"{resource}", method, {**query_params, "page": i}, sideloads, json=json
+        page_numbers = iter(range(2, last_page + 1))
+
+        def _fetch_page_request():
+            page_number = next(page_numbers)
+
+            async def fetch_page():
+                return await self._fetch_page(
+                    f"{resource}",
+                    method,
+                    {**query_params, "page": page_number},
+                    sideloads,
+                    json=json,
                 )
-            )
-            for i in range(2, last_page + 1)
+
+            return fetch_page
+
+        # initiate max_in_flight_requests requests in parallel
+        pending_request_tasks = [
+            *self._run_tasks(_fetch_page_request, self.max_in_flight_requests),
```

Comment:
This code feels quite complex. Looks like `asyncio` has a primitive called [asyncio.Semaphore](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore) for this. Can we use it? :pray: There might be smt even better but this is what I found :wink:

Reply:
@asgeirrr Refactored to use `asyncio.Semaphore`. Could you have another look?